fdbserver/storageserver.actor.cpp

```diff
@@ -8572,6 +8576,11 @@ ACTOR Future<Void> fetchKeys(StorageServer* data, AddingShard* shard) {
 					lastError = e;
 				}
 			}
+			if (lastError.code() == error_code_storage_replica_comparison_error) {
+				Version knownCommitted = data->knownCommittedVersion.get();
+				wait(data->version.whenAtLeast(knownCommitted));
```

Comment:
I think the logic is slightly different. It's the `data->knownCommittedVersion.get()` should be higher than last `fetchVersion`. I.e., previously `fetchVersion` may peek the data that later were rolled back. So we retry after last `fetchVersion` has become committed.

Reply:
> So we retry after last fetchVersion has become committed.
But there is no guarantee that that version would get committed? It could be rolled out too?

Reply:
> It's the data->knownCommittedVersion.get() should be higher than last fetchVersion

So fetching at "knownCommittedVersion.get()" (which is what this code is doing) should guarantee no inconsistencies? Essentially, we need to find a committed version and "knownCommittedVersion.get()" is a committed version? I am not totally sure that "knownCommittedVersion.get()" would be greater than the previous "fetchVersion" in all cases though.

Reply:
Yes, when we retry with a new `fetchVersion`, there is no guarantee the version would be committed.

Thinking in another way: when there is a rolled back version, the cluster must go through a recovery, which is usually the uncommon case. So it should be fine to retry a few times. Since recovery broadcast the rollback version to all SSes, all ongoing data movement will be rolled back, so there is no data inconsistency.

Assuming there is no rollback, then data inconsistencies among replica replies should be extremely rare, because any occurrence means a data corruption event. So we don't necessarily need to find a committed version. It's just that committed version guarantees that no rolled back data are read. However, the `fetchVersion` should be almost always fine to use, the chance of being wrong should usually be less than 0.1% (assuming 99.9% availability).

Reply:
I get the point. Are you saying the code should be like this?

```
if (lastError.code() == error_code_storage_replica_comparison_error) {
                                wait(data->knownCommittedVersion.whenAtLeast(fetchVersion);
				fetchVersion = data->knownCommittedVersion.get();
}
```

Reply:
I think the code is like this:
```
if (lastError.code() == error_code_storage_replica_comparison_error) {
      // Make sure last fetchVersion is already committed. I.e., in the case where
      // last time a replica's reply could contain rolled back version (due to recovery),
      // we retry with a new fetchVersion that is hopefully not going to be rolled back.
      wait(data->knownCommittedVersion.whenAtLeast(lastFetchVersion);
}
```
I'm thinking we don't need to change `fetchVersion` in this block, for the reason explained in the code comment.

Reply:
Makes sense. Maybe we should move this code to the catch-block (then the code would pick the right fetch version on the next iteration)? The fetch version gets assigned by the time this code currently runs and so I think it could pick up a stale fetch version again if we don't explicitly set that here. That is, the logic would like this: assign fetch version and do a wait on "data->knownCommittedVersion" exceeding the previous fetch version, so the fetch version could be stale?

Reply:
Yes. Moving this to the catch block is better.

Reply:
Done.